packages/beacon-node/src/execution/engine/http.ts

```diff
@@ -204,58 +206,56 @@ export class ExecutionEngineHttp implements IExecutionEngine {
     parentBlockRoot?: Root,
     executionRequests?: ExecutionRequests
   ): Promise<ExecutePayloadResponse> {
-    const method =
-      ForkSeq[fork] >= ForkSeq.electra
-        ? "engine_newPayloadV4"
-        : ForkSeq[fork] >= ForkSeq.deneb
-          ? "engine_newPayloadV3"
-          : ForkSeq[fork] >= ForkSeq.capella
-            ? "engine_newPayloadV2"
-            : "engine_newPayloadV1";
+    const method = newPayloadMethodByFork(fork);
 
     const serializedExecutionPayload = serializeExecutionPayload(fork, executionPayload);
 
     let engineRequest: EngineRequest;
-    if (ForkSeq[fork] >= ForkSeq.deneb) {
-      if (versionedHashes === undefined) {
-        throw Error(`versionedHashes required in notifyNewPayload for fork=${fork}`);
-      }
-      if (parentBlockRoot === undefined) {
-        throw Error(`parentBlockRoot required in notifyNewPayload for fork=${fork}`);
-      }
 
-      const serializedVersionedHashes = serializeVersionedHashes(versionedHashes);
-      const parentBeaconBlockRoot = serializeBeaconBlockRoot(parentBlockRoot);
-
-      if (ForkSeq[fork] >= ForkSeq.electra) {
-        if (executionRequests === undefined) {
-          throw Error(`executionRequests required in notifyNewPayload for fork=${fork}`);
+    switch (method) {
+      case "engine_newPayloadV1":
+      case "engine_newPayloadV2":
+        engineRequest = {
+          method,
+          params: [serializedExecutionPayload],
+          methodOpts: notifyNewPayloadOpts,
+        };
+        break;
+      case "engine_newPayloadV3":
+        if (versionedHashes === undefined || parentBlockRoot === undefined) {
+          throw Error(
+            `versionedHashes and parentBlockRoot required in notifyNewPayload for fork=${fork} versionedHashes=${versionedHashes} parentBlockRoot=${parentBlockRoot}`
+          );
         }
-        const serializedExecutionRequests = serializeExecutionRequests(executionRequests);
         engineRequest = {
-          method: "engine_newPayloadV4",
+          method,
           params: [
             serializedExecutionPayload,
-            serializedVersionedHashes,
-            parentBeaconBlockRoot,
-            serializedExecutionRequests,
+            serializeVersionedHashes(versionedHashes),
+            serializeBeaconBlockRoot(parentBlockRoot),
           ],
           methodOpts: notifyNewPayloadOpts,
         };
-      } else {
+        break;
+      case "engine_newPayloadV4":
+        if (versionedHashes === undefined || parentBlockRoot === undefined || executionRequests === undefined) {
```

Comment:
this will get out of hand now as we likely add a new property each fork, before the nesting kinda made this part a bit cleaner

I was thinking we might be able to do something like this

```ts
function assertNewPayloadParams<T>(
  fork: string,
  params: {[K in keyof T]: T[K] | undefined}
): asserts params is {[K in keyof T]: Exclude<T[K], undefined>} {
  for (const [key, value] of Object.entries(params)) {
    if (value === undefined) {
      throw new Error(`${key} is required in notifyNewPayload for fork=${fork}`);
    }
  }
}
```

and then just do
```ts
assertNewPayloadParams(fork, {versionedHashes, parentBlockRoot});
```

but typescript does not properly infer the type and still complains about it being undefined

Reply:
> I was thinking we might be able to do something like this
> 
> ```ts
> 
> ```

I think that's a good idea. Incorporated it into `generateSerializedNewPayloadParams()`